File: src/context.ts
import * as path from "jsr:@std/path";
import { Document, DOMParser } from "jsr:@b-fuze/deno-dom";
import {
    generate as cssGenerate,
    parse as cssParse,
    walk as cssWalk,
} from "npm:css-tree@2.3.1";
import { DO_NOT_FETCH_HOSTNAMES } from "./config.ts";

const USER_AGENT = "cohost-dl/1.0";
const MAX_FILE_NAME_LENGTH_UTF8 = 250;

export function splitTooLongFileName(filePath: string): string {
    let dirname = path.dirname(filePath);
    let filename = path.basename(filePath);

    while (
        new TextEncoder().encode(filename).byteLength >
            MAX_FILE_NAME_LENGTH_UTF8
    ) {
        let firstBit = "";
        let rest = filename;
        while (
            new TextEncoder().encode(firstBit).byteLength <
                MAX_FILE_NAME_LENGTH_UTF8
        ) {
            const item = rest[Symbol.iterator]().next().value;
            firstBit += item;
            rest = rest.substring(item.length);
        }

        dirname = path.join(dirname, firstBit);
        filename = rest;
    }

    return path.join(dirname, filename);
}

const NORMAL_FILE_EXTENSIONS = {
    // image formats
    "apng": "image/apng",
    "avif": "image/avif",
    "bmp": "image/bmp",
    "gif": "image/gif",
    "heic": "image/heic",
    "heif": "image/heif",
    "ico": "image/x-icon",
    "jpeg": "image/jpeg",
    "jpg": "image/jpeg",
    "jfif": "image/jpeg",
    "jxl": "image/jxl",
    "png": "image/png",
    "svg": ["image/svg+xml", "image/svg"],
    "tif": "image/tiff",
    "tiff": "image/tiff",
    "webp": "image/webp",

    // av formats
    "flac": "audio/flac",
    "ogg": ["audio/ogg", "video/ogg", "application/ogg"],
    "opus": "audio/opus",
    "mp3": "audio/mpeg",
    "mp4": ["audio/mp4", "video/mp4"],
    "m4a": ["audio/mp4", "video/mp4"],
    "wav": ["audio/wav", "audio/vnd.wave", "audio/wave", "audio/x-wav"],

    // other resources
    "css": "text/css",
    "js": ["application/javascript", "text/javascript"],
    "mjs": ["application/javascript", "text/javascript"],
    "json": ["application/json", "text/json"],
    "map": [] as string[],
    "woff": ["font/woff"],
    "woff2": ["font/woff2"],
};

function doesResourceFilePathProbablyNeedAFileExtension(
    filePath: string,
): boolean {
    const fileExtension = path.extname(filePath).toLowerCase().replace(
        /^[.]/,
        "",
    );
    return !(fileExtension in NORMAL_FILE_EXTENSIONS);
}

function resourceFileExtensionForContentType(
    contentType: string,
): string | null {
    const baseContentType = contentType.split(";")[0];

    for (const [ext, types] of Object.entries(NORMAL_FILE_EXTENSIONS)) {
        if (
            (Array.isArray(types) && types.includes(baseContentType)) ||
            types === baseContentType
        ) return ext;
    }
    return null;
}

export const POST_URL_REGEX = /^https:\/\/cohost[.]org\/([^\/]+)\/post\/(\d+)-/;

const CACHED_CONTENT_TYPES_FILE_PATH = "~headers.json";

export function encodeFilePathURI(path: string): string {
    // encodeURI preserves ?search, which we don’t want
    return path.split('/').map(encodeURIComponent).join('/');
}

export class CohostContext {
    /** cookie header */
    cookie: string;

    /** output directory for files */
    rootDir: string;

    cachedContentTypes: Record<string, Record<string, string>> = {};

    constructor(cookie: string, rootDir: string) {
        this.cookie = cookie;
        this.rootDir = rootDir;
    }

    async init() {
        try {
            this.cachedContentTypes = (await this.readJson(
                CACHED_CONTENT_TYPES_FILE_PATH,
            )) as typeof this.cachedContentTypes;
        } catch {
            // not important
        }
    }

    async finalize() {
        await this.flushCachedContentTypes();
    }

    getCachedContentType(url: string): string | null {
        return this.cachedContentTypes[url]?.["content-type"] ?? null;
    }

    setCachedContentType(url: string, contentType: string) {
        this.cachedContentTypes[url] = { "content-type": contentType };
        this.scheduleFlushCachedContentTypes();
    }

    scheduledFlushContentTypes: ReturnType<typeof setTimeout> | null = null;
    scheduleFlushCachedContentTypes() {
        if (this.scheduledFlushContentTypes) return;
        this.scheduledFlushContentTypes = setTimeout(() => {
            const _ = this.flushCachedContentTypes();
        }, 5000);
    }

    async flushCachedContentTypes() {
        if (this.scheduledFlushContentTypes) {
            clearTimeout(this.scheduledFlushContentTypes);
        }
        this.scheduledFlushContentTypes = null;

        try {
            await this.write(
                CACHED_CONTENT_TYPES_FILE_PATH,
                JSON.stringify(this.cachedContentTypes),
            );
        } catch (error) {
            console.error("failed to write content types cache!");
            console.error(error);
        }
    }

    /** Performs a GET request */
    async get(url: string): Promise<Response> {
        console.log(`GET ${url}`);
        const res = await fetch(url, {
            headers: {
                "user-agent": USER_AGENT,
                ...(new URL(url).host === "cohost.org"
                    ? { cookie: this.cookie }
                    : {}),
            },
        });

        if (!res.ok) {
            throw new Error(await res.text());
        }

        return res;
    }

    /** Beware when using this method with external files, because it may not return the final filePath */
    propsForResourceURL(
        urlString: string,
    ): {
        fetch: string;
        filePath: string;
        canFail?: boolean;
        skipFileExtCheck?: boolean;
    } | null {
        const url = new URL(urlString);
        if (
            url.hostname === "staging.cohostcdn.org" &&
            url.pathname.match(/^\/[a-z]+\//)
        ) {
            return {
                fetch: `https://staging.cohostcdn.org${url.pathname}`,
                filePath: `rc${decodeURIComponent(url.pathname)}`,
                skipFileExtCheck: true,
            };
        } else if (url.hostname === "cohost.org") {
            return {
                fetch: urlString,
                filePath: decodeURIComponent(url.pathname.substring(1)), // no leading /
                skipFileExtCheck: true,
            };
        } else if (
            url.protocol === "https:" &&
            !DO_NOT_FETCH_HOSTNAMES.includes(url.hostname)
        ) {
            return {
                fetch: urlString,
                filePath: splitTooLongFileName(
                    `rc/external/${url.host}${url.pathname}${url.search}`,
                ),
                canFail: true,
            };
        } else {
            console.error(`IGNORING URL ${urlString}`);
        }

        return null;
    }

    pendingResources = new Map<string, Promise<string | null>>();

    /** Ensure a valid path on disk relative to the root directory */
    getCleanPath(filePath: string): string {
        if (Deno.build.os === "windows") {
            // replace illegal characters for windows paths
            const cleanFilePath = filePath.replace(/[?%*:|"<>]/g, "-");
            return path.join(this.rootDir, cleanFilePath);
        }

        return path.join(this.rootDir, filePath);
    }

    async hasFile(filePath: string): Promise<boolean> {
        const fullPath = this.getCleanPath(filePath);
        try {
            await Deno.stat(fullPath);
            return true;
        } catch {
            return false;
        }
    }

    getCachedFileForPostURL(url: string): Promise<string | null> {
        const match = url.match(POST_URL_REGEX);
        if (!match) return Promise.resolve(null);
        const [, projectHandle, id] = match;

        return this.getCachedFileForPost(projectHandle, id);
    }

    async getCachedFileForPost(projectHandle: string, id: string | number): Promise<string | null> {
        const projectDir = path.join(this.getCleanPath(projectHandle), "post");
        try {
            for await (const item of Deno.readDir(projectDir)) {
                if (
                    item.name.startsWith(id + "-") &&
                    item.name.endsWith(".html")
                ) {
                    return `${projectHandle}/post/${item.name}`;
                }
            }
        } catch {
            // readDir failed - probably because the directory doesn't exist
        }

        return null;
    }

    readText(filePath: string): Promise<string> {
        const fullPath = this.getCleanPath(filePath);
        return Deno.readTextFile(fullPath);
    }

    async readJson(filePath: string): Promise<object> {
        return JSON.parse(await this.readText(filePath));
    }

    async write(filePath: string, data: string | Uint8Array) {
        const fullPath = this.getCleanPath(filePath);

        await Deno.mkdir(path.dirname(fullPath), { recursive: true });
        await Deno.writeFile(
            fullPath,
            typeof data === "string" ? new TextEncoder().encode(data) : data,
        );
    }

    async loadResourceToFile(urlString: string): Promise<string | null> {
        const props = this.propsForResourceURL(urlString);
        if (!props) return null;

        const needsFileExtension = !props.skipFileExtCheck &&
            doesResourceFilePathProbablyNeedAFileExtension(props.filePath);

        let forceFailStat = false;
        let filePathWithExt = props.filePath;

        if (needsFileExtension) {
            const cachedContentType = this.getCachedContentType(props.fetch);
            if (cachedContentType === null) {
                // from an older version of this script & probably broken. we need to reload this one
                forceFailStat = true;
            } else {
                const ext = resourceFileExtensionForContentType(
                    cachedContentType,
                );
                if (ext !== null) filePathWithExt = props.filePath + "." + ext;
            }
        }

        const fullPath = this.getCleanPath(filePathWithExt);

        try {
            if (forceFailStat) throw new Error("missing extension");

            await Deno.stat(fullPath);
            return filePathWithExt;
        } catch {
            const pending = this.pendingResources.get(urlString);
            if (pending) return pending;

            const pending2 = (async () => {
                let res: Response;
                try {
                    res = await this.get(props.fetch);
                } catch (err) {
                    if (props.canFail) {
                        console.error(`FAILED: GET ${props.fetch}`);
                        return null;
                    } else {
                        throw err;
                    }
                }

                const contentTypeHeader = res.headers.get("content-type");
                if (contentTypeHeader) {
                    this.setCachedContentType(props.fetch, contentTypeHeader);
                }

                if (needsFileExtension && contentTypeHeader) {
                    const ext = resourceFileExtensionForContentType(
                        contentTypeHeader,
                    );
                    if (ext !== null) {
                        filePathWithExt = props.filePath + "." + ext;
                    }
                }

                const baseContentType = contentTypeHeader?.split(";")?.[0];

                let data: string | Uint8Array;
                if (baseContentType === "text/css") {
                    data = await this.processCss(
                        urlString,
                        props.filePath,
                        await res.text(),
                    );
                } else {
                    data = new Uint8Array(await res.arrayBuffer());
                }

                await this.write(filePathWithExt, data);

                return filePathWithExt;
            })();
            this.pendingResources.set(urlString, pending2);
            return pending2;
        }
    }

    async processCss(
        urlString: string,
        filePath: string,
        contents: string,
    ): Promise<string> {
        const tree = cssParse(contents);
        const nodes: { value: string }[] = [];
        cssWalk(tree, (node: { type: string; value: string }) => {
            if (node.type === "Url") {
                nodes.push(node);
            }
        });

        const toRootDir =
            path.dirname(filePath).split("/").map(() => "..").join("/") + "/";

        await Promise.all(nodes.map(async (node) => {
            const resolved = new URL(node.value, urlString);
            if (resolved.protocol !== "https:") return;

            const filePath = await this.loadResourceToFile(resolved.toString());
            if (filePath) node.value = encodeFilePathURI(toRootDir + filePath);
        }));

        return cssGenerate(tree);
    }

    async getDocument(url: string, orCachedPath?: string): Promise<Document> {
        let htmlString: string | null = null;
        if (orCachedPath) {
            try {
                htmlString = await this.readText(orCachedPath);
            } catch {
                // probably doesn't exist
            }
        }

        if (htmlString === null) {
            htmlString = await this.get(url).then((res) => res.text());
        }

        return new DOMParser().parseFromString(htmlString || "", "text/html");
    }
}

File: db/src/context.rs
use crate::data::Database;
use crate::dl::CurrentStateV1;
use anyhow::{anyhow, Context};
use diesel::SqliteConnection;
use reqwest::{Client, IntoUrl, StatusCode, Url};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashSet;
use std::fs;
use std::io::Write;
use std::ops::Deref;
use std::path::{Path, PathBuf};
use std::time::Duration;
use tempfile::NamedTempFile;
use thiserror::Error;
use tokio::sync::Mutex;
use tokio::time::sleep;

pub const USER_AGENT: &str = "cohost-dl/2.0";
const MAX_FILE_NAME_LENGTH_UTF8: usize = 250;

pub struct CohostContext {
    cookie: String,
    client: Client,
    pub root_dir: PathBuf,
    temp_dir: PathBuf,
    pub do_not_fetch_domains: HashSet<String>,
    pub(crate) db: Database,
}

struct ResourceUrlProps {
    fetch: Url,
    file_path: PathBuf,
    can_fail: bool,
    skip_file_ext_check: bool,
}

#[derive(Debug, Error)]
pub enum GetError {
    #[error(transparent)]
    Url(reqwest::Error),
    #[error("{0} not found: {1}")]
    NotFound(Url, String),
    #[error("{0} {1}: {2}")]
    OtherStatus(Url, StatusCode, String),
    #[error("GET {0}: {1}")]
    Req(Url, reqwest::Error),
    #[error("{0:?}")]
    Other(anyhow::Error),
}

#[derive(Debug, Error)]
pub enum LoadResError {
    #[error(transparent)]
    Get(#[from] GetError),
    #[error("{0:?}")]
    Unknown(anyhow::Error),
}

impl GetError {
    pub fn is_recoverable(&self) -> bool {
        match self {
            GetError::OtherStatus(
                _,
                StatusCode::UNAUTHORIZED
                | StatusCode::FORBIDDEN
                | StatusCode::METHOD_NOT_ALLOWED
                | StatusCode::GONE,
                _,
            ) => false,
            GetError::NotFound(..) => false,
            GetError::Url(..) => false,
            _ => true,
        }
    }
}

impl LoadResError {
    pub fn is_recoverable(&self) -> bool {
        match self {
            LoadResError::Get(err) => err.is_recoverable(),
            LoadResError::Unknown(_) => true,
        }
    }
}

impl From<diesel::result::Error> for LoadResError {
    fn from(value: diesel::result::Error) -> Self {
        Self::Unknown(value.into())
    }
}

impl From<std::io::Error> for LoadResError {
    fn from(value: std::io::Error) -> Self {
        Self::Unknown(value.into())
    }
}

pub const MAX_RETRIES: usize = 10;

impl Deref for CohostContext {
    type Target = Database;

    fn deref(&self) -> &Self::Target {
        &self.db
    }
}

impl CohostContext {
    pub fn new(
        cookie: String,
        req_timeout: Duration,
        root_dir: PathBuf,
        db: SqliteConnection,
    ) -> Self {
        let client = Client::builder()
            .user_agent(USER_AGENT)
            .timeout(req_timeout)
            .build()
            .expect("failed to create client");

        let temp_dir = root_dir.join("tmp");
        let _ = fs::create_dir_all(&temp_dir);

        // we have to canonicalize these so that we get \\?\ file paths on Windows.
        // this allows us to use long file paths
        let root_dir = root_dir
            .canonicalize()
            .expect("could not canonicalize root directory");
        let temp_dir = temp_dir
            .canonicalize()
            .expect("could not canonicalize temp directory");

        CohostContext {
            cookie,
            client,
            root_dir,
            temp_dir,
            do_not_fetch_domains: Default::default(),
            db: Database::new(db),
        }
    }

    pub async fn get_text(&self, url: impl IntoUrl) -> Result<String, GetError> {
        let url = url.into_url().map_err(GetError::Url)?;
        let mut tries = 0;

        loop {
            if tries > 0 {
                let wait = 1.8_f64.powf(tries as f64) - 1.;
                info!(
                    "try {}: waiting for {wait:.02}s before continuing to be polite",
                    tries + 1
                );
                sleep(Duration::from_secs_f64(wait)).await;
            }

            tries += 1;
            trace!("GET {url}");

            let mut req = self.client.get(url.clone());

            if url.domain() == Some("cohost.org") {
                req = req.header("cookie", &self.cookie);
            }

            let res = req.send().await.map_err(|e| GetError::Req(url.clone(), e));

            let res = match res {
                Ok(res) => res,
                Err(e) if tries < MAX_RETRIES => {
                    error!("{e}. trying again (try {}/{MAX_RETRIES})", tries + 1);
                    continue;
                }
                e => e?,
            };

            let status = res.status();
            let text = res.text().await.map_err(|e| GetError::Req(url.clone(), e));

            let text = match text {
                Ok(text) => text,
                Err(e) if tries < MAX_RETRIES => {
                    error!("{e}. trying again (try {}/{MAX_RETRIES})", tries + 1);
                    continue;
                }
                e => e?,
            };

            if status.is_success() {
                return Ok(text);
            } else if status == StatusCode::NOT_FOUND {
                return Err(GetError::NotFound(url, text));
            } else if tries < MAX_RETRIES {
                error!(
                    "{status}: {text}. trying again (try {}/{MAX_RETRIES})",
                    tries + 1
                );
            } else {
                return Err(GetError::OtherStatus(url, status, text));
            }
        }
    }

    pub async fn get_json<T>(&self, url: impl IntoUrl) -> Result<T, GetError>
    where
        T: for<'a> Deserialize<'a> + 'static,
    {
        let url = url.into_url().map_err(GetError::Url)?;
        let text = self.get_text(url.clone()).await?;

        Ok(serde_json::from_str(&text)
            .map_err(|e| {
                let line = e.line().saturating_sub(1);
                let col = e.column().saturating_sub(1);

                let mut line_start = 0;
                for _ in 0..line {
                    let next_line_break = text[line_start..]
                        .char_indices()
                        .find(|(_, c)| *c == '\n')
                        .map(|(i, _)| i);
                    if let Some(next_line_break) = next_line_break {
                        line_start = next_line_break + 1;
                    }
                }

                let slice_center = line_start + col;
                let mut slice_start = slice_center;
                let mut slice_end = slice_center;

                for i in 0..4 {
                    let maybe_start = slice_center - 300 + i;
                    if text.is_char_boundary(maybe_start) {
                        slice_start = maybe_start;
                        break;
                    }
                }
                for i in 0..4 {
                    let maybe_end = (slice_center + 300).min(text.len()) + i;
                    if text.is_char_boundary(maybe_end) {
                        slice_end = maybe_end;
                        break;
                    }
                }

                let e = anyhow::Error::from(e);
                e.context(format!("excerpt: {}", &text[slice_start..slice_end]))
            })
            .map_err(GetError::Other)?)
    }

    pub async fn trpc_query<A, T>(&self, query: &str, input: Option<A>) -> Result<T, GetError>
    where
        A: Serialize,
        T: for<'a> Deserialize<'a> + 'static,
    {
        let mut url = Url::parse(&format!("https://cohost.org/api/v1/trpc/{query}")).unwrap();

        if let Some(input) = input {
            url.query_pairs_mut().append_pair(
                "input",
                &serde_json::to_string(&input).map_err(|e| GetError::Other(e.into()))?,
            );
        }

        #[derive(Deserialize)]
        enum TrpcResult<T> {
            #[serde(rename = "result")]
            Result { data: T },
            #[serde(rename = "error")]
            Error {
                code: i64,
                data: serde_json::Value,
                message: String,
            },
        }

        let result = self.get_json::<TrpcResult<T>>(url).await?;
        match result {
            TrpcResult::Result { data } => Ok(data),
            TrpcResult::Error {
                code,
                data,
                message,
            } => Err(GetError::Other(anyhow!(
                "TRPC error {code} / {data:?}: {message}"
            ))),
        }
    }

    pub async fn get_file(&self, url: impl IntoUrl) -> Result<reqwest::Response, GetError> {
        let url = url.into_url().map_err(GetError::Url)?;
        trace!("GET {url}");

        let res = self
            .client
            .get(url.clone())
            .send()
            .await
            .map_err(|e| GetError::Req(url.clone(), e))?;

        let status = res.status();
        if status.is_success() {
            Ok(res)
        } else {
            let err = res
                .text()
                .await
                .map_err(|e| GetError::Req(url.clone(), e))?;

            let mut err_slice_end = err.len().min(500);
            for i in 0..4 {
                if err.is_char_boundary(err_slice_end + i) {
                    err_slice_end = err_slice_end + i;
                    break;
                }
            }

            let err = &err[..err_slice_end];

            match status {
                StatusCode::NOT_FOUND => Err(GetError::NotFound(url, err.to_string())),
                status => Err(GetError::OtherStatus(url, status, err.to_string())),
            }
        }
    }

    fn props_for_resource_url(&self, url: &Url) -> anyhow::Result<Option<ResourceUrlProps>> {
        if url.domain() == Some("staging.cohostcdn.org")
            && url
                .path_segments()
                .unwrap()
                .next()
                .map_or(false, |seg| seg.chars().all(|c| c.is_ascii_alphabetic()))
        {
            let mut file_path = self.root_dir.clone();
            file_path.push("rc");
            for seg in url.path_segments().unwrap() {
                file_path.push(urlencoding::decode(seg)?.to_string());
            }

            Ok(Some(ResourceUrlProps {
                fetch: Url::parse(&format!("https://staging.cohostcdn.org/{}", url.path()))?,
                file_path,
                can_fail: false,
                skip_file_ext_check: true,
            }))
        } else if url.domain() == Some("cohost.org") {
            let mut file_path = self.root_dir.clone();
            for seg in url.path_segments().unwrap() {
                if seg.is_empty() {
                    continue;
                }
                file_path.push(urlencoding::decode(seg)?.to_string());
            }

            Ok(Some(ResourceUrlProps {
                fetch: url.clone(),
                file_path,
                can_fail: false,
                skip_file_ext_check: true,
            }))
        } else if url.scheme() == "https"
            && !url
                .domain()
                .map_or(true, |d| self.do_not_fetch_domains.contains(d))
        {
            let mut file_path = self.root_dir.clone();
            file_path.push("rc");
            file_path.push("external");
            file_path.push(url.host_str().unwrap_or_default());

            let mut additional_path = url.path().to_string();
            if let Some(query) = url.query() {
                additional_path.push('?');
                additional_path.push_str(query);
            }
            if let Some(frag) = url.fragment() {
                additional_path.push('#');
                additional_path.push_str(frag);
            }
            if additional_path.starts_with('/') {
                additional_path.remove(0);
            }
            if additional_path.ends_with('/') {
                additional_path.pop();
            }

            if additional_path.is_empty() {
                additional_path.push('_');
            }

            if additional_path.len() > 1536 {
                // probably too long
                let mut hasher = Sha256::new();
                hasher.update(additional_path);
                let result = hasher.finalize();
                additional_path = format!("(hash)_{}", hex::encode(result));
            }

            for seg in additional_path.split('/') {
                let seg = if cfg!(target_os = "windows") {
                    seg.chars()
                        .map(|c| match c {
                            '?' | '%' | '*' | ':' | '|' | '"' | '<' | '>' => '-',
                            c => c,
                        })
                        .collect()
                } else {
                    seg.to_string()
                };

                if seg.len() > MAX_FILE_NAME_LENGTH_UTF8 {
                    let mut buf = String::new();
                    for c in seg.chars() {
                        buf.push(c);
                        if buf.len() >= MAX_FILE_NAME_LENGTH_UTF8 {
                            file_path.push(&buf);
                            buf.clear();
                        }
                    }
                    if !buf.is_empty() {
                        file_path.push(buf);
                    }
                } else {
                    file_path.push(seg);
                }
            }

            Ok(Some(ResourceUrlProps {
                fetch: url.clone(),
                file_path,
                can_fail: true,
                skip_file_ext_check: false,
            }))
        } else {
            trace!("ignoring URL {url}");
            Ok(None)
        }
    }

    fn add_content_type_ext(file_path: PathBuf, content_type: &str, should_warn: bool) -> PathBuf {
        if let Some(ext) = resource_file_extension_for_content_type(content_type) {
            if let Some(file_name) = file_path.file_name() {
                let mut file_name = file_name.to_os_string();
                // JS implementation was *specifically* appending, so we're not using with_extension
                file_name.push(".");
                file_name.push(ext);

                let mut file_path = file_path;
                file_path.pop();
                file_path.push(file_name);
                file_path
            } else {
                if should_warn {
                    warn!(
                        "could not fix resource extension for {}: no file path?",
                        file_path.display()
                    );
                }
                file_path
            }
        } else {
            if should_warn {
                warn!(
                    "did not add missing file extension for {} because of unknown content type {:?}",
                    file_path.display(),
                    content_type
                );
            }
            file_path
        }
    }

    /// Returns the file path where the resource is supposed to be stored at.
    ///
    /// This exists because an older version stored them at the wrong path.
    pub async fn get_intended_resource_file_path(
        &self,
        url: &Url,
    ) -> anyhow::Result<Option<PathBuf>> {
        let Some(props) = self.props_for_resource_url(url)? else {
            return Ok(None);
        };

        let needs_file_extension = !props.skip_file_ext_check
            && does_resource_probably_need_a_file_extension(&props.file_path);

        if needs_file_extension {
            let content_type = self.get_res_content_type(&props.fetch).await?;
            Ok(Some(Self::add_content_type_ext(
                props.file_path,
                &content_type.unwrap_or_default(),
                false,
            )))
        } else {
            Ok(Some(props.file_path))
        }
    }

    /// Loads a resource to a file.
    /// Returns file path (relative to out dir) or None if it shouldn't be loaded
    pub async fn load_resource_to_file(
        &self,
        url: &Url,
        state: &Mutex<CurrentStateV1>,
        loaded: Option<&mut bool>,
    ) -> Result<Option<PathBuf>, LoadResError> {
        if state.lock().await.failed_urls.contains(&url.to_string()) {
            return Ok(None);
        }

        let Some(props) = self
            .props_for_resource_url(url)
            .map_err(|e| LoadResError::Unknown(e.into()))?
        else {
            return Ok(None);
        };

        let needs_file_extension = !props.skip_file_ext_check
            && does_resource_probably_need_a_file_extension(&props.file_path);
        let mut needs_content_type = false;

        let file_path_with_ext = if needs_file_extension {
            let content_type = self
                .get_res_content_type(&props.fetch)
                .await
                .map_err(|e| LoadResError::Unknown(e.into()))?;

            if content_type.is_none() {
                needs_content_type = true;
            }

            Self::add_content_type_ext(
                props.file_path.clone(),
                &content_type.unwrap_or_default(),
                false,
            )
        } else {
            props.file_path.clone()
        };

        if !needs_content_type {
            if let Some(result) = self.get_url_file(&url).await? {
                return Ok(Some(result));
            }
        }

        if fs::exists(&file_path_with_ext)? && !needs_content_type {
            let result_file_path = file_path_with_ext
                .strip_prefix(&self.root_dir)
                .context("getting relative path")
                .map_err(|e| LoadResError::Unknown(e.into()))?
                .to_path_buf();

            return Ok(Some(result_file_path));
        }

        let mut res = match self.get_file(url.clone()).await {
            Ok(file) => file,
            Err(e) => {
                if props.can_fail || !e.is_recoverable() {
                    state.lock().await.failed_urls.push(url.to_string());
                }
                Err(LoadResError::Get(e))?
            }
        };

        let content_type = if let Some(content_type) = res
            .headers()
            .get("content-type")
            .and_then(|value| value.to_str().ok().map(|s| s.to_string()))
        {
            self.insert_res_content_type(&props.fetch, &content_type)
                .await
                .map_err(|e| LoadResError::Unknown(e.into()))?;
            content_type
        } else {
            self.insert_res_content_type(&props.fetch, "")
                .await
                .map_err(|e| LoadResError::Unknown(e.into()))?;
            String::new()
        };

        let file_path_with_ext = if needs_file_extension {
            Self::add_content_type_ext(props.file_path, &content_type, true)
        } else {
            props.file_path
        };

        let file = NamedTempFile::with_prefix_in("cohost-dl-res-", &self.temp_dir)
            .context("creating temporary file")
            .map_err(|e| LoadResError::Unknown(e.into()))?;

        while let Some(chunk) = res
            .chunk()
            .await
            .map_err(|e| LoadResError::Unknown(e.into()))?
        {
            file.as_file().write_all(&chunk)?;
        }

        let result_file_path = file_path_with_ext
            .strip_prefix(&self.root_dir)
            .context("getting relative path")
            .map_err(|e| LoadResError::Unknown(e.into()))?
            .to_path_buf();

        let mut file_path_dir = file_path_with_ext.clone();
        file_path_dir.pop();
        fs::create_dir_all(file_path_dir)?;

        file.persist(&file_path_with_ext)
            .with_context(|| format!("moving resource to {}", file_path_with_ext.display()))
            .map_err(|e| LoadResError::Unknown(e.into()))?;

        if let Some(loaded) = loaded {
            *loaded = true;
        }

        Ok(Some(result_file_path))
    }
}

const KNOWN_FILE_EXTENSIONS: &[(&str, &[&str])] = &[
    // image formats
    ("apng", &["image/apng"]),
    ("avif", &["image/avif"]),
    ("bmp", &["image/bmp"]),
    ("gif", &["image/gif"]),
    ("heic", &["image/heic"]),
    ("heif", &["image/heif"]),
    ("ico", &["image/x-icon"]),
    ("jpeg", &["image/jpeg"]),
    ("jpg", &["image/jpeg"]),
    ("jfif", &["image/jpeg"]),
    ("jxl", &["image/jxl"]),
    ("png", &["image/png"]),
    ("svg", &["image/svg+xml", "image/svg"]),
    ("tif", &["image/tiff"]),
    ("tiff", &["image/tiff"]),
    ("webp", &["image/webp"]),
    // av formats
    ("flac", &["audio/flac"]),
    ("ogg", &["audio/ogg", "video/ogg", "application/ogg"]),
    ("opus", &["audio/opus"]),
    ("mp3", &["audio/mpeg"]),
    ("mp4", &["audio/mp4", "video/mp4"]),
    ("m4a", &["audio/mp4", "video/mp4"]),
    (
        "wav",
        &["audio/wav", "audio/vnd.wave", "audio/wave", "audio/x-wav"],
    ),
    // other resources
    ("css", &["text/css"]),
    ("js", &["application/javascript", "text/javascript"]),
    ("mjs", &["application/javascript", "text/javascript"]),
    ("json", &["application/json", "text/json"]),
    ("map", &[]),
    ("woff", &["font/woff"]),
    ("woff2", &["font/woff2"]),
];

fn does_resource_probably_need_a_file_extension(path: &Path) -> bool {
    if let Some(ext) = path.extension() {
        let ext = ext.to_ascii_lowercase();
        let Some(ext) = ext.to_str() else { return true };

        for (e, _) in KNOWN_FILE_EXTENSIONS {
            if *e == ext {
                return false;
            }
        }
        true
    } else {
        true
    }
}

fn resource_file_extension_for_content_type(content_type: &str) -> Option<&'static str> {
    let base_content_type = content_type.split(';').next()?;

    for (e, c) in KNOWN_FILE_EXTENSIONS {
        if c.contains(&base_content_type) {
            return Some(*e);
        }
    }

    None
}
